recent-topics: Update hash for view to be "#recent".

laurynmm opened the pull request:
Updates the hash used for the recent conversations view to be `"#recent"` instead of `"#recent_topics"`.

The old hash has a legacy implementation ~~with comments for it to be removed once servers cannot be directly updated from Zulip 5.x to the current version~~. This will need to be in place for old Welcome Bot messages in the future.

Fixes #23132. 

---

**Notes & Questions**:
- I added `window.location.replace("#recent")` in the `do_hashchange_normal` logic for `"#recent_topics"` to update the URL. This means the back/forward navigation is not broken by the redirect. Is there another way to update/rewrite the URL after the reload that would be better? And would it be worth it to add a node test for this, maybe in the "hash_interactions" test? 
- Because the hash changes a link in the potential messages from `Welcome Bot`is this really a temporary change? Or will we need some sort of redirect logic to always exist?
- I don't think we need to update the string saved on the server side for the default view because the default view hash is `"#"` or `""`. So I'm planning on doing that update/change in a separate PR.

---

**Screenshots and screen captures:**

<details>
<summary>Screenshot of new hash in URL</summary>

![Screenshot from 2022-10-24 14-31-06](https://user-images.githubusercontent.com/63245456/197536982-0b67260d-b916-4187-9d88-4471d91781c2.png)
</details>

---

**Self-review checklist**

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/version-control.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.

laurynmm commented on static/js/hash_util.js:
```diff
@@ -273,7 +273,7 @@ export function is_spectator_compatible(hash) {
         "",
         "narrow", // full #narrow hash handled in narrow.is_spectator_compatible
-        "recent_topics",
+        "recent",
```
I wasn't sure if we need to keep "recent_topics" here with the `TODO/compatibility` note as well.

timabbott replied:
It's probably a good idea; might not come up in practice but it's cheap to do.

laurynmm replied:
I think we do need to leave this for now, but it can be removed once one cannot upgrade directly from Zulip 5.x, because at that point the only situation with the "#recent_topics" has that we'll need to handle is from the Welcome Bot messages. The reload logic won't be needed.

I added a very short `TODO/compatibility` comment for this; adding a longer comment wasn't as readable, but we might want to try anyway?

timabbott replied:
This is why we generally prefer to avoid the inline comments, rather than comments on the line before what we're commenting. I rewrote this as a proper paragraph comment as part of merging this.

timabbott commented on static/js/hashchange.js:
```diff
@@ -180,7 +180,15 @@ function do_hashchange_normal(from_reload) {
+        // TODO/compatibility: In order to make the reload.js logic work
+        // after upgrades for self-hosted systems, treat "#recent_topics"
+        // as "#recent" until one can no longer directly upgrade from
+        // Zulip 5.x to the current version.
```
I'd move this comment inside the case block.

laurynmm replied:
Moved and updated to be a permanent comment about why we're keeping the logic for this old hash.

timabbott commented:
> Because the hash changes a link in the potential messages from Welcome Botis this really a temporary change? Or will we need some sort of redirect logic to always exist?

Good catch. The comment should be redone to just explain we need to handle the old URL forever because it appears in older welcome bot messages.

timabbott commented:
> This means the back/forward navigation is not broken by the redirect. Is there another way to update/rewrite the URL after the reload that would be better? And would it be worth it to add a node test for this, maybe in the "hash_interactions" test?

And yeah, that'd be a good place to add a test for this. I think this wants to call a method from `browser_history.js`; I'm not sure which one, but maybe this one? 

```
export function update_hash_internally_if_required(hash) {                                               
    if (window.location.hash !== hash) {                                                                 
        update(hash);                                                                                    
    }                                                                                                    
}                                                                                                        
```

laurynmm commented:
I think this wants to call a method from `browser_history.js`; I'm not sure which one, but maybe this one?
> 
> ```
> export function update_hash_internally_if_required(hash) {                                               
>     if (window.location.hash !== hash) {                                                                 
>         update(hash);                                                                                    
>     }                                                                                                    
> }                                                                                                        
> ```

So, `browser_history.update(new_hash)` updates `window.location.hash` and that will break the back/forward navigation. When you navigate back to `/#recent_topics`, the hash is reset to `#recent` meaning you lose any forward history in the browser session, and it's impossible to continue to navigate back past that point with keyboard shortcuts. You can manage it in the browser by holding the back button down for the dropdown menu.

When I was thinking this was just going to be an issue for folks during the update/upgrade, that seemed okay-ish, but not great. And since this is a permanent redirect for the Welcome Bot messages, I thought it better to use [`window.location.replace()`](https://developer.mozilla.org/en-US/docs/Web/API/Location/replace) since that means the user still can browse back/forward.

laurynmm commented:
**Updates**:
- Fixed for the comments from review notes above.
- Updated the commit message to close #23132 because the updates to the default view string and the frontend codebase can be done as follow-up tasks after some conversation in CZO.
- Expanded on my reasoning for not using an existing `browser_history.js` function for the URL redirect/rewrite in [this comment](https://github.com/zulip/zulip/pull/23329#issuecomment-1290276514).

timabbott commented:
> I thought it better to use [window.location.replace()](https://developer.mozilla.org/en-US/docs/Web/API/Location/replace)

Yeah, I think this is correct; we want this to be basically an invisible redirect. It seems possible that some of our other URL rewrites, like #settings => the first settings panel might benefit from using replace, but certainly that can be a follow-up thing to look at when convenient.

> I don't think we need to update the string saved on the server side for the default view because the default view hash is "#" or "". So I'm planning on doing that update/change in a separate PR.

Ahh, good point!

timabbott commented:
Merged after small tweaks to the comments, huge thanks for making this happen @laurynmm!
